nerfstudio/field_components/temporal_grid.py

```diff
@@ -0,0 +1,354 @@
+            new_index = [last_index_min + 1, last_index_max + 1] + prev
+            new_index = torch.as_tensor(new_index, dtype=torch.long)
+            new_permute = permute_base[:last_entry] + [0] + permute_base[last_entry:]
+            new_permute = torch.as_tensor(new_permute, dtype=torch.long)
+            index_list.append(torch.as_tensor(new_index, dtype=torch.long))
+            permute_list.append(torch.as_tensor(new_permute, dtype=torch.long))
+            row, mask_a, mask_b = to_sampling_index(index_list[-1], permute_list[-1], last_entry)
+            sampling_index.append(row)
+            index_a_mask.append(mask_a)
+            index_b_mask.append(mask_b)
+        self.register_buffer("index_list", torch.stack(index_list))
+        self.register_buffer("sampling_index", torch.stack(sampling_index))
+        # index_a_mask and index_b_mask are for inserting the combination weights
+        self.register_buffer("index_a_mask", torch.stack(index_a_mask))
+        self.register_buffer("index_b_mask", torch.stack(index_b_mask))
+
+    def __repr__(self):
+        """For debug and logging purpose."""
+        return (
+            f"GridEncoder: input_dim={self.input_dim} num_levels={self.num_levels} "
+            f"level_dim={self.level_dim} resolution={self.base_resolution} -> "
+            f"{int(round(self.base_resolution * self.per_level_scale ** (self.num_levels - 1)))} "
+            f"per_level_scale={self.per_level_scale:.4f} params={tuple(self.embeddings.shape)} "
+            f"gridtype={self.gridtype} align_corners={self.align_corners}"
+        )
+
+    def get_temporal_index(self, time: TensorType["bs"]) -> TensorType["bs", "temporal_index_dim"]:
+        """Convert the time into sampling index lists."""
+        row_idx_value = time * (len(self.sampling_index) - 1)
+        row_idx = row_idx_value.long()
+        row_idx[time == 1] = len(self.sampling_index) - 1
+        temporal_row_index = self.sampling_index[row_idx].float()
+        mask_a = self.index_a_mask[row_idx]
+        mask_b = self.index_b_mask[row_idx]
+        temporal_row_index[mask_a] = row_idx + 1 - row_idx_value
+        temporal_row_index[mask_b] = row_idx_value - row_idx
+        return temporal_row_index
+
+    def forward(self, xyz: TensorType["bs", "input_dim"], time: TensorType["bs", 1]) -> TensorType["bs", "output_dim"]:
+        """Forward and sampling feature vectors from the embedding.
+
+        Args:
+            xyz: input coords, should be in [0,1]
+            time: input time, should be in [0,1] with shape [bs, 1]
+        """
+        outputs = TemporalGridEncodeFunc.apply(
+            xyz,
+            self.get_temporal_index(time[:, 0]),
+            self.embeddings,
+            self.offsets,
+            self.per_level_scale,
+            self.base_resolution,
+            xyz.requires_grad,
+            self.gridtype_id,
+            self.align_corners,
+        )
+        return outputs
+
+    def get_temporal_tv_loss(self):
```

Comment:
missing `->` return

Reply:
I can pass the test with `TensorType[()]`, but I find that in another loss function `TensorType[0]` is used. (https://github.com/nerfstudio-project/nerfstudio/blob/a4b023b171eb162d2352b617b648765807ee61bc/nerfstudio/model_components/losses.py#L277)
`TensorType[0]` fails to pass the test
```
E           TypeError: the return value must be of type TensorType[0], got type TensorType[()] instead. 
```
Both of them are returning loss values, not sure which `TensorType` is the right one.

Reply:
If `TensorType[()]` works, thats fine with me.